src/Workspaces/Core/Portable/Workspace/Solution/DocumentState.cs

```diff
@@ -457,6 +449,97 @@ public DocumentState UpdateAnalyzerConfigSet(ValueSource<AnalyzerConfigSet> newA
+
+            // In the case where the config file didn't change between the last version
+            // and this version (in any significant way) we want to reuse the previous
+            // document to prevent excess unnecessary work.
+
+            return new AsyncLazy<TreeAndVersion>(
+                c => FullyParseTreeIfChangedByUpdatedConfigAsync(_treeSource, newAnalyzerConfigSet, filePath, c),
+                c => FullyParseTreeIfChangedByUpdatedConfig(_treeSource, newAnalyzerConfigSet, filePath, c),
+                cacheResult: true);
+        }
+
+        private ValueSource<TreeAndVersion> CreateLazyFullyParsedTree(ValueSource<AnalyzerConfigSet> newAnalyzerConfigSetSource, string filePath)
+            => CreateLazyFullyParsedTree(
+                TextAndVersionSource, Id.ProjectId, filePath, _options!, newAnalyzerConfigSetSource, _languageServices);
+
+        /// <summary>
+        /// Asynchronous version of <see cref="FullyParseTreeIfChangedByUpdatedConfig"/>.
+        /// Keep the code for both in sync.
+        /// </summary>
+        private async Task<TreeAndVersion> FullyParseTreeIfChangedByUpdatedConfigAsync(
+            ValueSource<TreeAndVersion> treeAndVersionSource, ValueSource<AnalyzerConfigSet> newAnalyzerConfigSetSource,
+            string filePath, CancellationToken cancellationToken)
+        {
+            var oldAnalyzerConfigSet = await _analyzerConfigSetSource.GetValueAsync(cancellationToken).ConfigureAwait(false);
+            var newAnalyzerConfigSet = await newAnalyzerConfigSetSource.GetValueAsync(cancellationToken).ConfigureAwait(false);
+
+            if (CanReuseExistingTree(filePath, oldAnalyzerConfigSet, newAnalyzerConfigSet))
+            {
+                // Config didn't change in a meaningful way.  Get the existing tree/version from
+                // before.
+                return await treeAndVersionSource.GetValueAsync(cancellationToken).ConfigureAwait(false);
+            }
+
+            // Config did change in a significant way.  Reparse the file fully.
+            return await CreateLazyFullyParsedTree(newAnalyzerConfigSetSource, filePath).GetValueAsync(cancellationToken).ConfigureAwait(false);
+        }
+
+        /// <summary>
+        /// Synchronous version of <see cref="FullyParseTreeIfChangedByUpdatedConfigAsync"/>.
+        /// Keep the code for both in sync.
+        /// </summary>
+        private TreeAndVersion FullyParseTreeIfChangedByUpdatedConfig(
+            ValueSource<TreeAndVersion> treeAndVersionSource, ValueSource<AnalyzerConfigSet> newAnalyzerConfigSetSource,
+            string filePath, CancellationToken cancellationToken)
+        {
+            var oldAnalyzerConfigSet = _analyzerConfigSetSource.GetValue(cancellationToken);
+            var newAnalyzerConfigSet = newAnalyzerConfigSetSource.GetValue(cancellationToken);
+
+            if (CanReuseExistingTree(filePath, oldAnalyzerConfigSet, newAnalyzerConfigSet))
```

Comment:
So, this change seems to definitely improve this scenario where we are re-using syntax trees if changes to editorconfig did not affect tree's diagnostic options. However, solution crawler would still kick of a full project level analysis for every edit in editorconfig, see [here](http://sourceroslyn.io/#Microsoft.CodeAnalysis.Features/SolutionCrawler/WorkCoordinator.cs,367). We presumably need another follow-up change to optimize this scenario for solution crawler as well.

Reply:
@mavasani i'm not sure how we would optimize that case.  If keys/values are changed in editorconfig literally every analyzer may need to rerun?  the only way around that is if we redesign the system such that analyzers have to be declarative about what they are impacted by.

I'm not sure what i can do here?

Reply:
I meant things like typing a comment can possibly be optimized

Reply:
I see. Do you want to gate on that optimization though?  Or can the rest of this proceed?

Reply:
Oh this comment is not specific to this change, which seems fine to me on it’s own . You probably want @jasonmalinowski to review it though.

Reply:
We can also avoid updating the **.editorconfig** model _live_, opting instead to apply the changes when the file is saved. I have no problem with this incremental improvement, but I wouldn't go so far as to say it _fixes_ the linked issue.